test/gp/kernel.jl

```diff
@@ -196,15 +196,41 @@ using LinearAlgebra
         end
         
         @timedtestset "NeuralKernelNetwork" begin
-            k1 = 0.5 * stretch(EQ(), 0.1)
-            k2 = 1.0 * stretch(PerEQ(1.0), 0.2)
-            prim_layer = Primitive(k1, k2)
+            @timedtestset "general test" begin
+                k1 = 0.5 * stretch(EQ(), 0.1)
+                k2 = 1.0 * stretch(PerEQ(1.0), 0.2)
+                prim_layer = Primitive(k1, k2)
 
-            lin = LinearLayer(2, 2)
-            nn = Chain(lin, product)
+                lin = LinearLayer(2, 2)
+                nn = Chain(lin, product)
 
-            nkn = NeuralKernelNetwork(prim_layer, nn)
-            differentiable_kernel_tests(nkn, ȳ, Ȳ, Ȳ_sq, X0, X1, X2; atol=1e-7, rtol=1e-7)
+                nkn = NeuralKernelNetwork(prim_layer, nn)
+                differentiable_kernel_tests(nkn, ȳ, Ȳ, Ȳ_sq, x0, x1, x2; atol=1e-7, rtol=1e-7)
+                differentiable_kernel_tests(nkn, ȳ, Ȳ, Ȳ_sq, X0, X1, X2; atol=1e-7, rtol=1e-7)
+            end
+            @timedtestset "kernel composition test" begin
+                k1 = 0.5 * stretch(EQ(), 0.1)
+                k2 = 1.0 * stretch(PerEQ(1.0), 0.2)
+                prim_layer = Primitive(k1, k2)
+                lin = LinearLayer(ones(1, 2))
+                nkn_add_kernel = NeuralKernelNetwork(prim_layer, lin)
+                nkn_prod_kernel = NeuralKernelNetwork(prim_layer, product)
+
+                sum_k = Stheno.softplus(1.0)*k1 + Stheno.softplus(1.0)*k2
+                prod_k = k1 * k2
```

Comment:
Yes, the `softplus` function is used to make sure matrix elements of linear transformation always be positive. See [here](https://github.com/HamletWantToCode/Stheno.jl/blob/1f85c9a20ff4074c81482832ed2b7b6e4117c0e5/src/gp/neural_kernel_network.jl#L10-L16).

Therefore, in `lin = LinearLayer(ones(1, 2))`,  the actual weights used for calculation aren't `ones(1, 2)`, it should be `softplus.(ones(1, 2))`.

Reply:
I think it would still be better to use randomised weights. Having the same values for each of the elements is still a slightly weird edge case. As a rule, I try to make any quantities used in tests as un-special as (unless specifically testing potentially worrisome corner cases) possible to avoid potential edge cases -- randomisation is a good way to achieve this.

Reply:
Random weights has now been used to this test case :)